Products/ZenRRD/zencommand.py

```diff
@@ -567,227 +618,282 @@ def _process_os_processes_results_in_sequence(
+                            "Failed to write to metric service  "
+                            "metadata=%s type=%s message=%s",
+                            dp.metadata, e.__class__.__name__, e,
                         )
-                    yield self._dataService.writeMetricWithMetadata(
-                        dp.dpName,
-                        value,
-                        dp.rrdType,
-                        min=dp.rrdMin,
-                        max=dp.rrdMax,
-                        threshEventData=threshData,
-                        metadata=dp.metadata,
-                    )
-                except Exception as e:
-                    log.exception(
-                        "Failed to write to metric service: %s %s",
-                        dp.metadata, e,
-                    )
 
-            eventList = results.events
-            exitCode = getattr(datasource.result, "exitCode", -1)
-            output = None
-            if not isinstance(datasource.result, Failure):
-                output = datasource.result.output.strip()
-            if exitCode == 0 and output:
-                # Ensure a CLEAR event is sent for any command that
-                # successfully completes
-                clearEvents = next(
-                    (
-                        event
-                        for event in eventList
-                        if event["severity"] == Clear
-                    ),
-                    None,
-                )
-                if not clearEvents:
-                    msg = "Datasource %s command completed successfully" % (
-                        datasource.name
-                    )
-                    event = self._makeCmdEvent(datasource, msg, severity=Clear)
-                    eventList.append(event)
-
-            # Send accumulated events
-            for event in eventList:
-                self._eventService.sendEvent(event, device=self._devId)
-        doTask_end = datetime.now()
-        duration = doTask_end - self._doTask_start
-        if duration > timedelta(seconds=self.interval):
-            log.warn(
-                "Collection for %s took %s seconds; "
-                "cycle interval is %s seconds.",
-                self.configId, duration.total_seconds(), self.interval,
-            )
-        else:
-            log.debug(
-                "Collection time for %s was %s seconds; "
-                "cycle interval is %s seconds.",
-                self.configId, duration.total_seconds(), self.interval,
+                if response.output.strip():
+                    # Ensure a CLEAR event is sent for any command that
+                    # successfully completes.  This means non-empty output
+                    # (and an exit code == 0).
+                    hasClearEvent = any((
+                        (event["severity"] == Clear)
+                        for event in results.events
+                    ))
+                    if not hasClearEvent:
```

Comment:
in the sequence of:
1. command failure
2. timeout
3. successful command
you will clear the timeout but not the command failure